charmhelpers/contrib/openstack/context.py

```diff
@@ -2181,9 +2181,64 @@ def __call__(self):
 class HostInfoContext(OSContextGenerator):
     """Context to provide host information."""
 
+    def _get_use_fqdn_hint(self):
+        """Get hint from local unit KV store for whether FQDN should be used
+
+        Depending on the workload a charm manages, the use of FQDN vs.
+        shortname may be a deploy-time decision, i.e. behavour can not change
+        on post deploy charm upgrade or configuration change.
+
+        The hint is passed on as a flag in the context to allow the decision to
+        be made in the Jinja2 configuration template.
+
+        :returns: True or False
+        :rtype: bool
+        """
+        db = kv()
+        return db.get('openstack-hostinfocontext-use-fqdn-hint', False)
+
+    def _get_canonical_name(self, name=None):
+        """Get the official FQDN of the host
+
+        The implementation of ``socket.getfqdn()`` in the standard Python
+        library does not exhaust all methods of getting the official name
+        of a host ref Python issue https://bugs.python.org/issue5004
+
+        This function mimics the behaviour of a call to ``hostname -f`` to
+        get the official FQDN but returns an empty string if it is
+        unsuccessful.
+
+        :param name: Shortname to get FQDN on
+        :type name: Optional[str]
+        :returns: The official FQDN for host or empty string ('')
+        :rtype: str
+        """
+        name = name or socket.gethostname()
+        fqdn = ''
+
+        if six.PY2:
+            exc = socket.error
+        else:
+            exc = OSError
+
+        try:
+            addrs = socket.getaddrinfo(
+                name, None, 0, socket.SOCK_DGRAM, 0, socket.AI_CANONNAME)
+        except exc:
+            pass
+        else:
+            for addr in addrs:
+                if addr[3]:
```

Comment:
This is not inside a try/except block, isn't it safer to check the length before trying to access it?

Reply:
Normally I would agree with you, but the library call is well [documented](https://docs.python.org/3/library/socket.html#socket.getaddrinfo) and defined, so I do not see an opportunity for it to deviate from that in this case.